README
NAME
    File::BSED - Search/Replace in Binary Files.

VERSION
    This document describes File::BSED version 0.6

SYNOPSIS
        use File::BSED qw(binary_search_replace binary_file_matches);

        # search/replace
        my $matches = binary_search_replace({
            search  => 'ff3cea3f0013',
            replace => 'b801000000c3',
            infile  => '/bin/ls',
            outfile => '/tmp/ls.patched'
        });
        if ($matches == -1) {
            my $error_string = File::BSED->errtostr();
            die "Error: $error_string\n";
        }
        print "Replaced $matches time(s)\n";

        # search
        my $matches = binary_file_matches('ff3cea3f0013', '/bin/ls');
        print "Matched $matches time(s)\n";

        # replace 'gnu' to 'bsd'
        my $hex_gnu = File::BSED::string_to_hexstring('gnu');
        my $hex_bsd = File::BSED::string_to_hexstring('bsd');
        my $matches = binary_search_replace({
            search  => $hex_gnu,
            replace => $hex_bsd,
            infile  => '/bin/ls',
            outfile => 'ls.bsd',
        });
        # [...]

DESCRIPTION
    This is a perl-binding to "libgbsed", a binary stream editor.

    "gbsed" lets you search and replace binary data in binary files by using
    hex values in text strings as search patterns. You can also use wildcard
    matches with "??", which will match any wide byte.

    These are all valid search strings:

        search => "0xffc300193ab2f63a";
        search => "0xff??00??3ab2f??a";
        search => "FF??00??3AB2F??A";

    while these are not:

        search => "the quick brown fox"; # only hex, no text. you would have to
                                         # convert the text to hex first.
        search => "0xff?c33ab3?accc";    # no nybbles only wide bytes. (?? not ?).

SUBROUTINES/METHODS
  CLASS METHODS
   "binary_search_replace(\%arguments)"
    Search and replace in a binary file. Valid arguments are:

    search
        The pattern to search for. Must be a string with hex digits.

    replace
        What to replace with. Must be a string with hex digits. (Probably a
        very good idea that it is the same length as the search pattern).

    infile
        The filename of the file to search in.

    outfile
        The filename of the file to save changes to.

    minmatch
        Need at least $minmatch matches before any work.

    maxmatch
        Stop after $maxmatch matches.

    "bsed" returns the number of times the search string was found in the
    file, or -1 if an error occurred. When an error occurs the error number
    can be found with "errno" and you can get a description of the error
    with "errtostr".

    Example:

        my $number_of_matches = binary_search_replace({
            search  => '0xff',
            replace => '0x00',
            
        infile  => '/bin/ls',
            outfile => '/bin/ls.patched',
        });

        if ($number_of_matches = -1) {
            die sprintf("ERROR: %s\n", File::BSED->errtostr());
        }
        print "Replaced $number_of_matches time(s).\n";

   "gbsed(\%arguments)"
    Alias to "binary_search_replace" for backward compatibility.

   "binary_file_matches($search_pattern, $input_file)"
    Return the number of times $search_pattern is found in binary file.
    Search pattern must be a string of hex digits.

    Example:

        my $number_of_matches = binary_file_matches('0xfeedface', '/bin/ls');
        print "Matched $number_of_matches time(s)\n";

   "string_to_hexstring($text)"
    Convert text to a string of ASCII hex values.

  ERROR HANDLING
   "errtostr"
    This function returns a string describing what happened. if an error has
    occurred with either "binary_search_matches" or "binary_file_matches"
    (they return -1 on error).

   "errno"
    This function returns the error number if an error has occurred with
    either "binary_search_replace" or "binary_file_matches" (they return -1
    on error).

    Use "errtostr" instead.

  WARNINGS
   "warnings()"
    Returns an array ref to any warnings raised.

    Example:

        my @warnings = @{ File::BSED::warnings() };
        for my $warning (@warnings) {
            warn "Warning: $warning\n";
        }

DIAGNOSTICS
CONFIGURATION AND ENVIRONMENT
    This module requires no configuration file or environment variables.

DEPENDENCIES
    *   A working C compiler.

INCOMPATIBILITIES
    None known.

BUGS AND LIMITATIONS
    No bugs have been reported.

    Please report any bugs or feature requests to
    "bug-file-bsed@rt.cpan.org", or through the web interface at
    <http://rt.cpan.org>.

SEE ALSO
    *   plbsed

AUTHOR
    by Ask Solem, "ask@0x61736b.net".

ACKNOWLEDGEMENTS
    Dave Dykstra "dwdbsed@drdykstra.us". for "bsed" the original program,

    *0xfeedface* for the wildcards patch.

LICENSE AND COPYRIGHT
    Copyright (c), 2007 Ask Solem "ask@0x61736b.net".

  SEE
    perlartistic perlgpl

    All rights reserved.

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself, either Perl version 5.8.6 or, at
    your option, any later version of Perl 5 you may have available.

DISCLAIMER OF WARRANTY
    BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
    FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
    OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
    PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
    EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
    WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
    ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
    YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
    NECESSARY SERVICING, REPAIR, OR CORRECTION.

    IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
    WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
    REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE LIABLE
    TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL, OR
    CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE THE
    SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
    RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
    FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
    SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
    DAMAGES.



Changes
Revision history for Perl extension File-BSED.

0.6  Tue Jul 17 17:23:02 2007
    - uses new version of libgbsed (0.6)
      this means that now the output file will preserve the executable flags
      from the input file.
    - fixed a broken preprocessor directive when PERL_MALLOC was active
      in libgbsed.h
    - removed unused function _hexstr2bin from BSED.xs
    - tried to use goto() for some of the aliases in lib/File/BSED.pm, but
      this obviosly didn't work very well. Reset to function calls instead.

0.5  Sun Jul 15 22:38:45 2007
    - uses new version of libgbsed.

0.4  Fri Jul 13 17:33:09 2007
    - libgbsed.c: Forgot to close files and flush buffers.
    - libgbsed.c: Global variables removed.

0.3  Tue Jul 10 18:43:22 2007
    - bin/plbsed: We didn't import binary_file_matches
    - Documented wildcard matching and minmatch+maxmatch.
    - Changed Dave Dykstra's email from dwd@bell-labs.com to
      dwdbsed@drdykstra.us.

0.2  Tue Jul 10 17:33:53 2007
    - Removed t/XSExample.t as this was a file left over from h2xs,
      and made our tests fail.
    - Minor documentation changes.

0.1  Mon Jul 09 16:04:06 2007
	- Initial version.



BSED.c
/*
 * This file was generated automatically by ExtUtils::ParseXS version 2.18 from the
 * contents of BSED.xs. Do not edit this file, edit BSED.xs instead.
 *
 *	ANY CHANGES MADE HERE WILL BE LOST! 
 *
 */

#line 1 "BSED.xs"
#include "EXTERN.h"
#include "perl.h"
#include "XSUB.h"

#include "ppport.h"

#include <stdio.h>
#include "bsedlib.h"

extern int isb_errno;

#ifndef PERL_UNUSED_VAR
#  define PERL_UNUSED_VAR(var) if (0) var = var
#endif

#line 26 "BSED.c"

XS(XS_File__BSED__binary_file_matches); /* prototype to pass -Wmissing-prototypes */
XS(XS_File__BSED__binary_file_matches)
{
#ifdef dVAR
    dVAR; dXSARGS;
#else
    dXSARGS;
#endif
    if (items != 2)
       Perl_croak(aTHX_ "Usage: %s(%s)", "File::BSED::_binary_file_matches", "search, infile");
    PERL_UNUSED_VAR(cv); /* -W */
    {
	unsigned char *	search = (unsigned char *)SvPV_nolen(ST(0));
	unsigned char *	infile = (unsigned char *)SvPV_nolen(ST(1));
	int	RETVAL;
	dXSTARG;
#line 21 "BSED.xs"
        unsigned char *replace;
        unsigned char *outfile;
        int min  =  1;
        int max  = -1;
        int no_r =  1;
        replace  = NULL;
        outfile  = NULL;


    RETVAL = bsr(search, replace, infile, outfile, min, max, no_r);


#line 57 "BSED.c"
	XSprePUSH; PUSHi((IV)RETVAL);
    }
    XSRETURN(1);
}


XS(XS_File__BSED__bsed); /* prototype to pass -Wmissing-prototypes */
XS(XS_File__BSED__bsed)
{
#ifdef dVAR
    dVAR; dXSARGS;
#else
    dXSARGS;
#endif
    if (items != 6)
       Perl_croak(aTHX_ "Usage: %s(%s)", "File::BSED::_bsed", "search, replace, infile, outfile, min, max");
    PERL_UNUSED_VAR(cv); /* -W */
    {
	unsigned char *	search = (unsigned char *)SvPV_nolen(ST(0));
	unsigned char *	replace = (unsigned char *)SvPV_nolen(ST(1));
	unsigned char *	infile = (unsigned char *)SvPV_nolen(ST(2));
	unsigned char *	outfile = (unsigned char *)SvPV_nolen(ST(3));
	int	min = (int)SvIV(ST(4));
	int	max = (int)SvIV(ST(5));
	int	RETVAL;
	dXSTARG;
#line 46 "BSED.xs"
        int no_r =  0; 

    RETVAL = bsr(search, replace, infile, outfile, min, max, no_r);

#line 89 "BSED.c"
	XSprePUSH; PUSHi((IV)RETVAL);
    }
    XSRETURN(1);
}


XS(XS_File__BSED__isb_errno); /* prototype to pass -Wmissing-prototypes */
XS(XS_File__BSED__isb_errno)
{
#ifdef dVAR
    dVAR; dXSARGS;
#else
    dXSARGS;
#endif
    if (items != 0)
       Perl_croak(aTHX_ "Usage: %s(%s)", "File::BSED::_isb_errno", "");
    PERL_UNUSED_VAR(cv); /* -W */
    {
	int	RETVAL;
	dXSTARG;
#line 56 "BSED.xs"
        RETVAL = isb_errno;
#line 112 "BSED.c"
	XSprePUSH; PUSHi((IV)RETVAL);
    }
    XSRETURN(1);
}


XS(XS_File__BSED__set_isb_errno); /* prototype to pass -Wmissing-prototypes */
XS(XS_File__BSED__set_isb_errno)
{
#ifdef dVAR
    dVAR; dXSARGS;
#else
    dXSARGS;
#endif
    if (items != 1)
       Perl_croak(aTHX_ "Usage: %s(%s)", "File::BSED::_set_isb_errno", "new_errno");
    PERL_UNUSED_VAR(cv); /* -W */
    {
	int	new_errno = (int)SvIV(ST(0));
#line 65 "BSED.xs"
        isb_errno = new_errno;
#line 134 "BSED.c"
    }
    XSRETURN_EMPTY;
}


XS(XS_File__BSED__errtostr); /* prototype to pass -Wmissing-prototypes */
XS(XS_File__BSED__errtostr)
{
#ifdef dVAR
    dVAR; dXSARGS;
#else
    dXSARGS;
#endif
    if (items != 1)
       Perl_croak(aTHX_ "Usage: %s(%s)", "File::BSED::_errtostr", "isb_errno_val");
    PERL_UNUSED_VAR(cv); /* -W */
    {
	int	isb_errno_val = (int)SvIV(ST(0));
	SV *	RETVAL;
#line 73 "BSED.xs"
    const char *isb_errstr;

    isb_errstr = isb_errtostr(isb_errno_val);
    if (isb_errstr == NULL) {
        RETVAL = &PL_sv_undef;
    }
    else {
        RETVAL = newSVpvn((char *)isb_errstr, (STRLEN)strlen((char *)isb_errstr));
    }

#line 165 "BSED.c"
	ST(0) = RETVAL;
	sv_2mortal(ST(0));
    }
    XSRETURN(1);
}

#ifdef __cplusplus
extern "C"
#endif
XS(boot_File__BSED); /* prototype to pass -Wmissing-prototypes */
XS(boot_File__BSED)
{
#ifdef dVAR
    dVAR; dXSARGS;
#else
    dXSARGS;
#endif
    char* file = __FILE__;

    PERL_UNUSED_VAR(cv); /* -W */
    PERL_UNUSED_VAR(items); /* -W */
    XS_VERSION_BOOTCHECK ;

        newXS("File::BSED::_binary_file_matches", XS_File__BSED__binary_file_matches, file);
        newXS("File::BSED::_bsed", XS_File__BSED__bsed, file);
        newXS("File::BSED::_isb_errno", XS_File__BSED__isb_errno, file);
        newXS("File::BSED::_set_isb_errno", XS_File__BSED__set_isb_errno, file);
        newXS("File::BSED::_errtostr", XS_File__BSED__errtostr, file);
    if (PL_unitcheckav)
         call_list(PL_scopestack_ix, PL_unitcheckav);
    XSRETURN_YES;
}



bsedlib.h
/*
* $Id: bsedlib.h,v 1.2 2007/07/13 09:56:26 ask Exp $
* $Source: /opt/CVS/File-BSED/bsedlib.h,v $
* $Author: ask $
* $HeadURL$
* $Revision: 1.2 $
* $Date: 2007/07/13 09:56:26 $
*/

/*
 * bsed - binary stream editor.  Written Feb 1987 by David W. Dykstra
 * Copyright (C) 1987-2002 by Dave Dykstra and Lucent Technologies
 *
 * So what's with the copyright notice if this is GPL?
 * Anyway, monsterously hacked by feedface 2006
 * Now you must enter hex strings (without preceding 0x)
 * but you can do wildcard search/replace using ?? as a placeholder
 * ie ?? in the search string represents any byte
 * in the replace string ?? means do not replace that byte
 *
 */

/* If given a search=replace string, both input and output file names 	*/
/*   must be present; else only the input file name must be present.	*/
/* Both input and output file names can be "-" to signify standard 	*/
/*   input or standard output respectively.				*/

typedef unsigned char UCHAR;
int bsr(UCHAR *, UCHAR *, UCHAR *, UCHAR *, int, int, int);
int convert(register unsigned char *, register unsigned char *);
unsigned char* dump(unsigned char *str, register int len);
const char* isb_errtostr(int);

#define ISB_ERROR                -1
#define ISB_ESEARCH_TOO_LONG     0x1
#define ISB_ENULL_SEARCH         0x2
#define ISB_EREPLACE_TOO_LONG    0x3
#define ISB_ENULL_REPLACE        0x4
#define ISB_EMISSING_INPUT       0x5
#define ISB_EMISSING_OUTPUT      0x6

/*
# Local Variables:
#   indent-level: 4
#   fill-column: 78
# End:
# vim: expandtab tabstop=4 shiftwidth=4 shiftround
*/


libgbsed.h
/*
*
*    Copyright (C) 2007 Ask Solem <ask@0x61736b.net>
*
*    This file is part of gbsed
*
*    gbsed is free software; you can redistribute it and/or modify
*    it under the terms of the GNU General Public License as published by
*    the Free Software Foundation; either version 3 of the License, or
*    (at your option) any later version.
*
*    gbsed is distributed in the hope that it will be useful,
*    but WITHOUT ANY WARRANTY; without even the implied warranty of
*    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*    GNU General Public License for more details.
*
*    You should have received a copy of the GNU General Public License
*    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

/*
* $Id: libgbsed.h,v 1.4 2007/07/17 15:26:35 ask Exp $
* $Source: /opt/CVS/File-BSED/libgbsed.h,v $
* $Author: ask $
* $HeadURL$
* $Revision: 1.4 $
* $Date: 2007/07/17 15:26:35 $
*/

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif /* HAVE_CONFIG_H */

#ifndef _LIBGBSED_H_
#define _LIBGBSED_H_


/*           Constants             */

#define GBSED_ERROR                 -1
#define GBSED_NO_MATCH             0x0

#define GBSED_ESEARCH_TOO_LONG     0x1
#define GBSED_ENULL_SEARCH         0x2
#define GBSED_EREPLACE_TOO_LONG    0x3
#define GBSED_ENULL_REPLACE        0x4
#define GBSED_EMISSING_INPUT       0x5
#define GBSED_EMISSING_OUTPUT      0x6
#define GBSED_EINVALID_CHAR        0x7 
#define GBSED_ENIBBLE_NOT_BYTE     0x8 
#define GBSED_EOPEN_INFILE         0x9
#define GBSED_EOPEN_OUTFILE        0xa
#define GBSED_ENOMEM               0xb
#define GBSED_EMINMAX_BALANCE      0xc
#define GBSED_ENOSTAT_FDES          0xd

#define GBSED_WBALANCE             0x1

#define GBSED_MAXMATCH_NO_LIMIT     -1

/*           Types                 */

typedef unsigned char UCHAR;

struct gbsed_arguments {
    char *search;
    char *replace;
    char *infilename;
    char *outfilename;
    int  minmatch;
    int  maxmatch;
};
typedef struct gbsed_arguments  GBSEDargs;

struct fgbsed_arguments {
    char *search;
    char *replace;
    FILE *infile;
    FILE *outfile;
    int   minmatch;
    int   maxmatch;
};
typedef struct fgbsed_arguments fGBSEDargs;

/*           Public functions.      */

const char*
gbsed_version(void);

int
gbsed_binary_search_replace(struct gbsed_arguments *);

int
gbsed_fbinary_search_replace(struct fgbsed_arguments *);

char *
gbsed_string2hexstring(char *orig);

const char*
gbsed_errtostr(int);

char*
gbsed_warntostr(int);


void *
_gbsed_alloczero(size_t,  size_t);




#ifdef PERL_MALLOC
#  include <EXTERN.h>
#  include <perl.h>
#  define _gbsed_alloc(pointer, add, type)   \
    (type *)Newxz(pointer, add, type)
#  define  _gbsed_realloc(pointer, add, type) \
    (type *)Renew(pointer, add, type)
#  define _gbsed_safefree(pointer)           \
    Safefree(pointer)
#else /* not PERL_MALLOC */
#define _gbsed_alloc(pointer, add, type)    \
    (type *)_gbsed_alloczero(add, sizeof(type))
#define _gbsed_realloc(pointer, add, type)  \
    (type *)realloc(pointer, add*sizeof(type))
#define _gbsed_safefree(pointer)            \
    free(pointer)
#endif /* PERL_MALLOC */

/*           Private functions       */

#ifdef LIBGBSED_PRIVATE

char *
_gbsed_remove_0x_from_str(char *);

UCHAR *
_gbsed_hexstr2bin(register UCHAR *, int *);

mode_t
_gbsed_preserve_execbit(FILE *file);

#else  /* not lIBGBSED_PRIVATE */
#  define LIBGBSED_PRIVATE 0
#endif /*     LIBGBSED_PRIVATE */

#endif /* _LIBGBSED_H_ */


/*
# Local Variables:
#   indent-level: 4
#   fill-column: 78
# End:
# vim: expandtab tabstop=4 shiftwidth=4 shiftround
*/


lib/File/BSED.pm
# $Id: BSED.pm,v 1.15 2007/07/17 15:26:37 ask Exp $
# $Source: /opt/CVS/File-BSED/lib/File/BSED.pm,v $
# $Author: ask $
# $HeadURL$
# $Revision: 1.15 $
# $Date: 2007/07/17 15:26:37 $
package File::BSED;
use strict;
use warnings;
use 5.006006;
use Exporter;
use Carp qw(croak);
use vars qw($VERSION @EXPORT_OK @ISA);

$VERSION   = 0.66;

@EXPORT_OK = qw(
    gbsed
    binary_search_replace
    binary_file_matches
    string_to_hexstring
);

use base qw(Exporter);

require XSLoader;
XSLoader::load(__PACKAGE__, $VERSION);

sub binary_search_replace {
    my ($args_ref) = @_;
    croak 'Argument to File::BSED::binary_search_replace must be hash reference'
        if !ref $args_ref || ref $args_ref ne 'HASH';

    my $search  = $args_ref->{search};
    my $replace = $args_ref->{replace};
    my $infile  = $args_ref->{infile};
    my $outfile = $args_ref->{outfile};

    my $min     = $args_ref->{minmatch} ||  1;
    my $max     = $args_ref->{maxmatch} || -1;


    return _gbsed($search, $replace, $infile, $outfile, $min, $max);
}

sub binary_file_matches {
    my ($search, $infile) = @_;

    return 0 if not $search;

    croak 'Missing filename argument to binary_file_matches()'
        if not defined $infile;

    return _binary_file_matches($search, $infile);
}

sub string_to_hexstring {
    my ($string) = @_;
    return if not $string;

    return _string_to_hexstring($string);
}

sub gbsed {
    goto &binary_search_replace;
}

sub errtostr {
    return _errtostr(_gbsed_errno());
}

sub errno {
    return _gbsed_errno();
}

sub warnings {
    return _warnings();
}

1;

__END__
=pod


=head1 NAME

File::BSED - Search/Replace in Binary Files.

=head1 VERSION

This document describes File::BSED version 0.66

=head1 SYNOPSIS

    use File::BSED qw(binary_search_replace binary_file_matches);

    # search/replace
    my $matches = binary_search_replace({
        search  => 'ff3cea3f0013',
        replace => 'b801000000c3',
        infile  => '/bin/ls',
        outfile => '/tmp/ls.patched'
    });
    if ($matches == -1) {
        my $error_string = File::BSED->errtostr();
        die "Error: $error_string\n";
    }
    print "Replaced $matches time(s)\n";

    # search
    my $matches = binary_file_matches('ff3cea3f0013', '/bin/ls');
    print "Matched $matches time(s)\n";

    # replace 'gnu' to 'bsd'
    my $hex_gnu = File::BSED::string_to_hexstring('gnu');
    my $hex_bsd = File::BSED::string_to_hexstring('bsd');
    my $matches = binary_search_replace({
        search  => $hex_gnu,
        replace => $hex_bsd,
        infile  => '/bin/ls',
        outfile => 'ls.bsd',
    });
    # [...]


=head1 DESCRIPTION

This is a perl-binding to C<libgbsed>, a binary stream editor.

C<gbsed> lets you search and replace binary data in binary files by using hex
values in text strings as search patterns. You can also use wildcard matches
with C<??>, which will match any wide byte.

These are all valid search strings:

    search => "0xffc300193ab2f63a";
    search => "0xff??00??3ab2f??a";
    search => "FF??00??3AB2F??A";

while these are not:

    search => "the quick brown fox"; # only hex, no text. you would have to
                                     # convert the text to hex first.
    search => "0xff?c33ab3?accc";    # no nybbles only wide bytes. (?? not ?).



=head1 SUBROUTINES/METHODS


=head2 CLASS METHODS 

=head3 C<binary_search_replace(\%arguments)>

Search and replace in a binary file.
Valid arguments are:

=over 4

=item search

The pattern to search for. Must be a string with hex digits.

=item replace

What to replace with. Must be a string with hex digits.
(Probably a very good idea that it is the same length as the search pattern).

=item infile

The filename of the file to search in.

=item outfile

The filename of the file to save changes to.

=item minmatch

Need at least C<$minmatch> matches before any work.

=item maxmatch

Stop after C<$maxmatch> matches.

=back

C<bsed> returns the number of times the search string was found
in the file, or C<-1> if an error occurred.
When an error occurs the error number can be found with C<errno>
and you can get a description of the error with C<errtostr>.

Example:

    my $number_of_matches = binary_search_replace({
        search  => '0xff',
        replace => '0x00',
        
        infile  => '/bin/ls',
        outfile => '/bin/ls.patched',
    });

    if ($number_of_matches = -1) {
        die sprintf("ERROR: %s\n", File::BSED->errtostr());
    }
    print "Replaced $number_of_matches time(s).\n";

=head3 C<gbsed(\%arguments)>

Alias to C<binary_search_replace> for backward compatibility.

=head3 C<binary_file_matches($search_pattern, $input_file)>

Return the number of times C<$search_pattern> is found in binary file.
Search pattern must be a string of hex digits.

Example:

    my $number_of_matches = binary_file_matches('0xfeedface', '/bin/ls');
    print "Matched $number_of_matches time(s)\n";

=head3 C<string_to_hexstring($text)>

Convert text to a string of ASCII hex values.

=head2 ERROR HANDLING

=head3 C<errtostr>

This function returns a string describing what happened.
if an error has occurred with either C<binary_search_matches> or C<binary_file_matches>
(they return -1 on error). 

=head3 C<errno>

This function returns the error number if an error has occurred with
either C<binary_search_replace> or C<binary_file_matches> (they return -1 on error). 

Use C<errtostr> instead.

=head2 WARNINGS

=head3 C<warnings()>

Returns an array ref to any warnings raised.

Example:

    my @warnings = @{ File::BSED::warnings() };
    for my $warning (@warnings) {
        warn "Warning: $warning\n";
    }
    
=head1 DIAGNOSTICS

=head1 CONFIGURATION AND ENVIRONMENT

This module requires no configuration file or environment variables.

=head1 DEPENDENCIES


=over 4

=item * A working C compiler.

=back


=head1 INCOMPATIBILITIES

None known.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests to
C<bug-file-bsed@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org>.

=head1 SEE ALSO

=over 4

=item * L<plbsed>

=back

=head1 AUTHOR

by Ask Solem, C<< ask@0x61736b.net >>.

=head1 ACKNOWLEDGEMENTS

Dave Dykstra C<< dwdbsed@drdykstra.us >>.
for C<bsed> the original program,

I<0xfeedface>
for the wildcards patch.

=head1 LICENSE AND COPYRIGHT

Copyright (c), 2007 Ask Solem C<< ask@0x61736b.net >>.

=head2 SEE

L<perlartistic> L<perlgpl>

All rights reserved.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.6 or,
at your option, any later version of Perl 5 you may have available.

=head1 DISCLAIMER OF WARRANTY

BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY FOR THE
SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN OTHERWISE
STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES PROVIDE THE
SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESSED OR IMPLIED,
INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
FITNESS FOR A PARTICULAR PURPOSE. THE ENTIRE RISK AS TO THE QUALITY AND
PERFORMANCE OF THE SOFTWARE IS WITH YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE,
YOU ASSUME THE COST OF ALL NECESSARY SERVICING, REPAIR, OR CORRECTION.

IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING WILL ANY
COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR REDISTRIBUTE THE
SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE LIABLE TO YOU FOR DAMAGES,
INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING
OUT OF THE USE OR INABILITY TO USE THE SOFTWARE (INCLUDING BUT NOT LIMITED TO
LOSS OF DATA OR DATA BEING RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR
THIRD PARTIES OR A FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER
SOFTWARE), EVEN IF SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE
POSSIBILITY OF SUCH DAMAGES.

=cut

# Local Variables:
#   mode: cperl
#   cperl-indent-level: 4
#   fill-column: 78
# End:
# vim: expandtab tabstop=4 shiftwidth=4 shiftround
